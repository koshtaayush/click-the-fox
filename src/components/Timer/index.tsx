//Library
import React from 'react'
import styled from 'styled-components'
import { useDispatch, useSelector } from 'react-redux'
import { useHistory } from 'react-router-dom'

//Utils
import { useInterval } from './../../utils/useInterval'

//Actions
import { addPlayerPerformance } from './../../actions/appActions/app.actions'

//Shared interfaces
import {
    playerPerformance
} from './../../typings/sharedInterfaces'

//Configs
import { TIMER_VALUE } from './../../config/constants.config'

interface Props {}

const Timer: React.FC<Props> = () => {
    const { currentScore, playerPlaying } = useSelector(
        (state: {
            app: {
                currentScore: number
                playerPlaying: string
            }
        }) => ({
            currentScore: state.app.currentScore,
            playerPlaying: state.app.playerPlaying,
        })
    )

    //Local state to maintain timer
    const [seconds, setSeconds] = React.useState<number>(TIMER_VALUE)

    const dispatch = useDispatch()
    const history = useHistory()

    /**
     * Hook for setInterval to avoid closure stale memory issue
     */
    useInterval(() => {
        if (seconds > 0) {
            setSeconds(seconds - 1)
        } else {
            handleMoveToScore()
            history.push('/score')
        }
    }, 1000)

    /**
     * Function to move user to score screen once timer has ended
     */
    const handleMoveToScore = () => {
        let data: playerPerformance = {
            name: playerPlaying,
            datePlayed: new Date(),
            score: currentScore,
        }
        dispatch(addPlayerPerformance(data))
    }

    return (
        <React.Fragment>
            <TimeContainer>Time Left: {seconds}</TimeContainer>
        </React.Fragment>
    )
}

export default Timer

const TimeContainer = styled.div``
