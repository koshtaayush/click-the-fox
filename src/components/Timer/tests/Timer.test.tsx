import React from 'react'
import { shallow } from 'enzyme'
import * as Redux from 'react-redux'

import Timer from './../index'


jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useHistory: () => ({
        push: jest.fn(),
    }),
}))

describe('Timer', () => {

    const defaultState = {
        app: {
            currentScore: 10,
            playerPlaying: 'Ayush'
        },
    }

    const setUpUseSelector = (state = {}) => {
        const localState = {
            app: { ...defaultState.app, ...state },
        }
        const setupState = { ...localState }
        const useSelectorMock = jest.spyOn(Redux, 'useSelector')
        return useSelectorMock.mockImplementation((callback) =>
            callback(setupState)
        )
    }

    const setUpUseDispatch = (state = {}) => {
        const useDispatchMock = jest.spyOn(Redux, 'useDispatch')
        const mockDispatch = jest.fn()
        return useDispatchMock.mockReturnValue(mockDispatch)
    }

    beforeEach(() => {
        setUpUseDispatch()
    })


    it('Should match the snapshot', () => {

        setUpUseSelector()

        const component = shallow(
            <Timer />
        )

        expect(component).toMatchSnapshot()
    })
})
