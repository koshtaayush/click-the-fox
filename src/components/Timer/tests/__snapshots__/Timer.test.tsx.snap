// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer Should match the snapshot 1`] = `
<Fragment>
  <styled.div>
    Time Left: 
    30
  </styled.div>
</Fragment>
`;
